library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java

```diff
@@ -435,17 +435,46 @@ public void put(Slice key, long timeBucket, Slice value)
      * Free memory up to the given windowId
      * This method will be called by another thread. Adding concurrency control to Stream would impact the performance.
      * This method only calculates the size of the memory that could be released and then sends free memory request to the operator thread
+     *
+     * We intend to manage memory by keyStream and valueStream. But the we can't avoid caller use other mechanism to manage memory.
+     * It is required to cleanup maps in this case.
      */
     @Override
     public long freeMemory(long windowId) throws IOException
     {
-      // calculate the size first and then send the release memory request. It could reduce the chance of conflict and increase the performance.
-      long size = keyStream.dataSizeUpToWindow(windowId) + valueStream.dataSizeUpToWindow(windowId);
+
+      long memoryFreed = 0;
+      Long clearWindowId;
+
+      while ((clearWindowId = committedData.floorKey(windowId)) != null) {
```

Comment:
each floorKey and remove call is O(log n) so the loop is going to be O(n log n). Why don't we just have a normal ordered iteration to make it O(n) instead?